src/compiler/crystal/tools/typer.cr

```diff
@@ -0,0 +1,612 @@
+
+        unless node.abstract?
+          format_nested_with_end body
+        end
+
+        @vars.pop
+        @inside_def -= 1
+
+        false
+      end
+
+      def visit(node : Crystal::Arg)
+        @last_arg_is_skip = false
+
+        restriction = node.restriction
+        default_value = node.default_value
+
+        if @inside_lib > 0
+          # This is the case of `fun foo(Char)`
+          if !@token.type.ident? && restriction
+            accept restriction
+            return false
+          end
+        end
+
+        if node.name.empty?
+          skip_space_or_newline
+        else
+          @vars.last.add(node.name)
+
+          at_skip = at_skip?
+
+          if !at_skip && node.external_name != node.name
+            if node.external_name.empty?
+              write "_"
+            elsif @token.type.delimiter_start?
+              accept Crystal::StringLiteral.new(node.external_name)
+            else
+              write @token.value
+            end
+            write " "
+            next_token_skip_space_or_newline
+          end
+
+          @last_arg_is_skip = at_skip?
+
+          write @token.value
+          next_token
+        end
+
+        if restriction
+          skip_space_or_newline
+          write_token " ", :OP_COLON, " "
+          skip_space_or_newline
+          accept restriction
+          # ===== BEGIN NEW CODE =====
+          # If the current arg doesn't have a restriction already and we have a signature, write in the type restriction
+        elsif (sig = @signatures[@current_def.try &.location.to_s || 0_u64]?) && sig.args[node.external_name]?
+          skip_space_or_newline
+          write " : #{sig.args[node.external_name]}"
+          @added_types = true
+          # ===== END NEW CODE =====
+        end
+
+        if default_value
+          # The default value might be a Proc with args, so
+          # we need to remember this and restore it later
+          old_last_arg_is_skip = @last_arg_is_skip
+
+          skip_space_or_newline
+
+          check_align = check_assign_length node
+          write_token " ", :OP_EQ, " "
+          before_column = @column
+          skip_space_or_newline
+          accept default_value
+          check_assign_align before_column, default_value if check_align
+
+          @last_arg_is_skip = old_last_arg_is_skip
+        end
+
+        # This is the case of an enum member
+        if @token.type.op_semicolon?
+          next_token
+          @lexer.skip_space
+          if @token.type.comment?
+            write_comment
+            @exp_needs_indent = true
+          else
+            write ";" if @token.type.const?
+            write " "
+            @exp_needs_indent = @token.type.newline?
+          end
+        end
+
+        false
+      end
+    end
+
+    # A visitor for defs, oddly enough.
+    #
+    # Walk through the AST and capture all references to Defs that match a def_locator
+    class DefVisitor < Crystal::Visitor
```

Comment:
DefVisitor, a crystal visitor for.... visiting defs. Used for capturing a list of all passed through defs that match a def_locator used when running the CLI command.